# Covid19 - Counties Dataset
### The novel coronavirus has had huge implications to the world around us and how we live our day-to-day lives. In this notebook I'd like to explore how the coronavirus has spread over time especially around where I live in Northern New Jersey. As this area borders NYC, the new epicenter of the virus New Jersey is not surprisingly the state with the second most confirmed cases at the time of this analysis.

#### It's important to be mindful that these are confirmed cases which may or may not be indicative of the actual infections. Availability of testing kits and hospitals that allow testing for covid19 are scarce and it's hard to say if some parts of America test less than other places. 

#### Throughout the notebook I'll be commenting my train of thought which should translate to why I go a certain path with the analysis. The data used here is provided through the New York Times. 

In [18]:
# import libraries
import pandas as pd
import numpy as np

# import data
data = pd.read_csv('dataset/us-counties.csv')

# preview the data
print('Number of data points:', data.shape[0])
data.head()

Number of data points: 30843


date     county       state     fips  cases  deaths
0  2020-01-21  Snohomish  Washington  53061.0      1       0
1  2020-01-22  Snohomish  Washington  53061.0      1       0
2  2020-01-23  Snohomish  Washington  53061.0      1       0
3  2020-01-24       Cook    Illinois  17031.0      1       0
4  2020-01-24  Snohomish  Washington  53061.0      1       0

In [23]:
# check if empty entries exist
data.isnull().sum()

date      0
county    0
state     0
cases     0
deaths    0
dtype: int64

There seems to be some blank entries for the column fips which is a unique code for each county. As we'll only be doing analysis on the county and states we can omit the fips column entirely for the purposes of this analysis.

In [21]:
# remove fips column
data.drop('fips', axis=1, inplace=True)

Let's run some explaratory analysis now we know this data is in a usable state.

In [31]:
# min/max dates in dataset
print('Min date:', min(data.date))
print('Max date:', max(data.date))
print('')

# number of states in dataset
print('Number of distinct states in dataset:', data.state.nunique())

Min date: 2020-01-21
Max date: 2020-04-03

Number of distinct states in dataset: 55


Um.. 55 distinct states?? We'll have to list them out and vett the list to see what's included. In the case there's typos in the state names we'll have to do additional data cleaning.

In [45]:
for state in sorted(data.state.unique()):
    print(state)

Alabama
Alaska
Arizona
Arkansas
California
Colorado
Connecticut
Delaware
District of Columbia
Florida
Georgia
Guam
Hawaii
Idaho
Illinois
Indiana
Iowa
Kansas
Kentucky
Louisiana
Maine
Maryland
Massachusetts
Michigan
Minnesota
Mississippi
Missouri
Montana
Nebraska
Nevada
New Hampshire
New Jersey
New Mexico
New York
North Carolina
North Dakota
Northern Mariana Islands
Ohio
Oklahoma
Oregon
Pennsylvania
Puerto Rico
Rhode Island
South Carolina
South Dakota
Tennessee
Texas
Utah
Vermont
Virgin Islands
Virginia
Washington
West Virginia
Wisconsin
Wyoming


District of Columbia, Guam, Northern Mariana Islands and Virgin Islands are not typical states, but we'll keep them in the dataset. Let's rank the number of cases and deaths for each of these states.

As we will running analysis on multiple levels of data we'll make a function for adding some useful statistics for the table.

In [71]:
def create_rate_cols(tdf):
    '''Adds death rate, cases percent of total and deaths percent of total columns to a dataframe.'''
    df = tdf.copy()
    sum_cases = sum(df.cases)
    sum_deaths = sum(df.deaths)
    
    # add columns
    df['death_rate'] = df.deaths / df.cases
    df['cases perc. of total'] = df.cases / sum_cases
    df['deaths perc. of total'] = df.deaths / sum_deaths
    
    cols = ['death_rate', 'cases perc. of total', 'deaths perc. of total']
    
    # convert decimals to percentage
    for col in cols:
        df[col] = df[col].astype(float).map('{:.2%}'.format)
    
    return df

In [73]:
# sum the cases and deaths per state
state_cd = data.groupby('state').sum()[['cases', 'deaths']]

# add stats columns
state_cd = create_rate_cols(state_cd)

# print dataframe sorted by death cases
state_cd.sort_values('cases', ascending=False)

cases  deaths death_rate cases perc. of total  \
state                                                                      
New York                  743198   12602      1.70%               40.61%   
New Jersey                170274    2762      1.62%                9.30%   
California                 89145    1847      2.07%                4.87%   
Michigan                   71954    2198      3.05%                3.93%   
Washington                 63368    3193      5.04%                3.46%   
Florida                    62627     902      1.44%                3.42%   
Massachusetts              61632     809      1.31%                3.37%   
Illinois                   57310     971      1.69%                3.13%   
Louisiana                  54795    2098      3.83%                2.99%   
Pennsylvania               45088     539      1.20%                2.46%   
Georgia                    38318    1230      3.21%                2.09%   
Texas                      35446     508      1.43%                1.94%   
Colorado                   29895     625      2.09%                1.63%   
Connecticut                26902     612      2.27%                1.47%   
Tennessee                  20943     170      0.81%                1.14%   
Ohio                       20878     456      2.18%                1.14%   
Indiana                    19069     491      2.57%                1.04%   
Maryland                   15871     179      1.13%                0.87%   
Wisconsin                  14576     255      1.75%                0.80%   
North Carolina             14176      83      0.59%                0.77%   
Virginia                   12151     184      1.51%                0.66%   
Missouri                   11968     181      1.51%                0.65%   
Arizona                    11647     179      1.54%                0.64%   
Nevada                     11214     236      2.10%                0.61%   
South Carolina             10876     218      2.00%                0.59%   
Alabama                    10042     140      1.39%                0.55%   
Mississippi                 9412     165      1.75%                0.51%   
Utah                        8778      39      0.44%                0.48%   
Oregon                      7086     191      2.70%                0.39%   
Minnesota                   6706      88      1.31%                0.37%   
Arkansas                    6527      89      1.36%                0.36%   
Oklahoma                    5657     202      3.57%                0.31%   
Kentucky                    5653     172      3.04%                0.31%   
District of Columbia        5314      82      1.54%                0.29%   
Idaho                       4934      62      1.26%                0.27%   
Rhode Island                4875      54      1.11%                0.27%   
Iowa                        4655      57      1.22%                0.25%   
Kansas                      4076     104      2.55%                0.22%   
New Hampshire               3666      33      0.90%                0.20%   
Maine                       3278      37      1.13%                0.18%   
New Mexico                  3250      39      1.20%                0.18%   
Delaware                    3098      68      2.19%                0.17%   
Vermont                     3048     142      4.66%                0.17%   
Nebraska                    2633      30      1.14%                0.14%   
Hawaii                      2330       7      0.30%                0.13%   
Puerto Rico                 1975      73      3.70%                0.11%   
Montana                     1922      26      1.35%                0.11%   
West Virginia               1516       8      0.53%                0.08%   
Alaska                      1296      14      1.08%                0.07%   
North Dakota                1269      18      1.42%                0.07%   
Wyoming                     1235       0      0.00%                0.07%   

New York has almost 41% of all coronavirus confirmed cases and New York and New Jersey combined have nearly half of all of America's infections. I live right on the border of NY and NJ which is the most heavily populated area in NJ so I assume most of the coronavirus cases are going to be in my county (which is Bergen). For NY I assume most of their cases is in NYC.

In [78]:
# create dataset of only New York and New Jersey
nyj = data[data.state.isin(['New York', 'New Jersey'])]
nj = data[data.state=='New Jersey']
ny = data[data.state=='New York']

# let's first look at the counties in New Jersey
nj_county = nj.groupby('county').sum().sort_values('cases', ascending=False)

# create rate columns
nj_county = create_rate_cols(nj_county)

# print dataframe
nj_county

cases  deaths death_rate cases perc. of total  \
county                                                      
Unknown     33043      85      0.26%               19.41%   
Bergen      28562     627      2.20%               16.77%   
Essex       16691     479      2.87%                9.80%   
Hudson      13842     186      1.34%                8.13%   
Union       12418     179      1.44%                7.29%   
Middlesex   11981     222      1.85%                7.04%   
Passaic     11078     125      1.13%                6.51%   
Monmouth    10750     206      1.92%                6.31%   
Ocean        9236     173      1.87%                5.42%   
Morris       7506     194      2.58%                4.41%   
Somerset     3821     118      3.09%                2.24%   
Mercer       2666      13      0.49%                1.57%   
Camden       2157      29      1.34%                1.27%   
Burlington   1944      45      2.31%                1.14%   
Sussex       1163      29      2.49%                0.68%   
Gloucester    979      13      1.33%                0.57%   
Hunterdon     917       0      0.00%                0.54%   
Warren        754      22      2.92%                0.44%   
Atlantic      330       2      0.61%                0.19%   
Cape May      174       0      0.00%                0.10%   
Cumberland    168      10      5.95%                0.10%   
Salem          94       5      5.32%                0.06%   

           deaths perc. of total  
county                            
Unknown                    3.08%  
Bergen                    22.70%  
Essex                     17.34%  
Hudson                     6.73%  
Union                      6.48%  
Middlesex                  8.04%  
Passaic                    4.53%  
Monmouth                   7.46%  
Ocean                      6.26%  
Morris                     7.02%  
Somerset                   4.27%  
Mercer                     0.47%  
Camden                     1.05%  
Burlington                 1.63%  
Sussex                     1.05%  
Gloucester                 0.47%  
Hunterdon                  0.00%  
Warren                     0.80%  
Atlantic                   0.07%  
Cape May                   0.00%  
Cumberland                 0.36%  
Salem                      0.18%

Looks like there's an unknown category for county which may be due to lack of information from the source. We'll remove the unknown county entries for now to not skew the percentages, but also be aware we're removing some of the data from the dataset.

In [81]:
# recreate dataframe with New Jersey counties
nj_county = nj.groupby('county').sum().sort_values('cases', ascending=False)

# remove unknown county label
nj_county = nj_county[nj_county.index!='Unknown']

# create rate columns
nj_county = create_rate_cols(nj_county)

# print dataframe
nj_county

cases  deaths death_rate cases perc. of total  \
county                                                      
Bergen      28562     627      2.20%               20.81%   
Essex       16691     479      2.87%               12.16%   
Hudson      13842     186      1.34%               10.09%   
Union       12418     179      1.44%                9.05%   
Middlesex   11981     222      1.85%                8.73%   
Passaic     11078     125      1.13%                8.07%   
Monmouth    10750     206      1.92%                7.83%   
Ocean        9236     173      1.87%                6.73%   
Morris       7506     194      2.58%                5.47%   
Somerset     3821     118      3.09%                2.78%   
Mercer       2666      13      0.49%                1.94%   
Camden       2157      29      1.34%                1.57%   
Burlington   1944      45      2.31%                1.42%   
Sussex       1163      29      2.49%                0.85%   
Gloucester    979      13      1.33%                0.71%   
Hunterdon     917       0      0.00%                0.67%   
Warren        754      22      2.92%                0.55%   
Atlantic      330       2      0.61%                0.24%   
Cape May      174       0      0.00%                0.13%   
Cumberland    168      10      5.95%                0.12%   
Salem          94       5      5.32%                0.07%   

           deaths perc. of total  
county                            
Bergen                    23.42%  
Essex                     17.89%  
Hudson                     6.95%  
Union                      6.69%  
Middlesex                  8.29%  
Passaic                    4.67%  
Monmouth                   7.70%  
Ocean                      6.46%  
Morris                     7.25%  
Somerset                   4.41%  
Mercer                     0.49%  
Camden                     1.08%  
Burlington                 1.68%  
Sussex                     1.08%  
Gloucester                 0.49%  
Hunterdon                  0.00%  
Warren                     0.82%  
Atlantic                   0.07%  
Cape May                   0.00%  
Cumberland                 0.37%  
Salem                      0.19%

So the first known county of most cases is Bergen county.. where I live. It is a little more spread out than I would've guessed however as I thought the percentage of cases in Bergen county would've been much more than 20.81% of total cases in New Jersey.

Let's now look at New York at their counties.

In [82]:
# recreate dataframe with New Jersey counties
ny_county = ny.groupby('county').sum().sort_values('cases', ascending=False)

# remove unknown county label
ny_county = ny_county[ny_county.index!='Unknown']

# create rate columns
ny_county = create_rate_cols(ny_county)

# print dataframe
ny_county

cases  deaths death_rate cases perc. of total  \
county                                                          
New York City  423482    9946      2.35%               56.98%   
Westchester    102386     253      0.25%               13.78%   
Nassau          82243     600      0.73%               11.07%   
Suffolk         62543     552      0.88%                8.42%   
Rockland        26828     195      0.73%                3.61%   
Orange          15349     102      0.66%                2.07%   
Erie             4547      89      1.96%                0.61%   
Dutchess         4532      38      0.84%                0.61%   
Monroe           2986      66      2.21%                0.40%   
Albany           2792      12      0.43%                0.38%   
Onondaga         1942       5      0.26%                0.26%   
Ulster           1887      14      0.74%                0.25%   
Putnam           1784       0      0.00%                0.24%   
Saratoga         1293       6      0.46%                0.17%   
Sullivan         1058      15      1.42%                0.14%   
Schenectady      1021      10      0.98%                0.14%   
Tompkins          610       0      0.00%                0.08%   
Rensselaer        590       2      0.34%                0.08%   
Niagara           465       0      0.00%                0.06%   
Madison           382       6      1.57%                0.05%   
Oneida            370       3      0.81%                0.05%   
Broome            335      34     10.15%                0.05%   
Columbia          304       8      2.63%                0.04%   
Steuben           254       2      0.79%                0.03%   
Ontario           226       0      0.00%                0.03%   
Clinton           197       0      0.00%                0.03%   
St. Lawrence      190       0      0.00%                0.03%   
Wayne             182       0      0.00%                0.02%   
Chenango          166       0      0.00%                0.02%   
Chemung           148       0      0.00%                0.02%   
Delaware          148       2      1.35%                0.02%   
Warren            144       0      0.00%                0.02%   
Greene            139       0      0.00%                0.02%   
Herkimer          137      10      7.30%                0.02%   
Oswego            130       6      4.62%                0.02%   
Otsego            125       3      2.40%                0.02%   
Livingston        112       2      1.79%                0.02%   
Wyoming           108       4      3.70%                0.01%   
Genesee            99       3      3.03%                0.01%   
Jefferson          95       0      0.00%                0.01%   
Washington         91       0      0.00%                0.01%   
Allegany           80       3      3.75%                0.01%   
Montgomery         80       0      0.00%                0.01%   
Cortland           70       0      0.00%                0.01%   
Essex              62       0      0.00%                0.01%   
Schoharie          59       0      0.00%                0.01%   
Tioga              58       0      0.00%                0.01%   
Franklin           56       0      0.00%                0.01%   
Chautauqua         54       3      5.56%                0.01%   
Orleans            43       0      0.00%                0.01%   
Cattaraugus        40       0      0.00%                0.01%   
Cayuga             35       0      0.00%                0.00%   
Hamilton           32       0      0.00%                0.00%   
Fulton             26       0      0.00%                0.00%   
Seneca             14       0      0.00%                0.00%   
Schuyler           14       0      0.00%                0.00%   
Lewis              14       0      0.00%                0.00%   
Yates               4       0      0.00%                0.00%   

              deaths perc. of total  
county                               
New York City                82.92%  
W

New York City has 57% of the cases in the state of New York and 423,482 confirmed cases which is a staggering number. A quick look at the census data provided by NYC open data says for 2020 there's approximately 8,550,971 people living in NYC. 

This means 4.95% of the people living in NYC today are confirmed to have covid19.